README.md
# Flamel

[//]: # (Badges)
[![GitHub Actions Build Status](https://github.com/alchemistry/flamel/workflows/CI/badge.svg)](https://github.com/alchemistry/flamel/actions?query=workflow%3ACI)
[![codecov](https://codecov.io/gh/alchemistry/flamel/branch/master/graph/badge.svg?token=NvoYnT9NSb)](https://codecov.io/gh/alchemistry/flamel)

The aim of the project is to develop a **command line interface (CLI) to 
[alchemlyb](https://github.com/alchemistry/alchemlyb)**, the well-tested and 
actively developed library for alchemical free energy calculations. It is 
supposed to [become the successor](https://github.com/alchemistry/alchemlyb/wiki/Roadmap#librarify-alchemical-analysis-functionality) 
of the now unsupported [alchemical-analysis](https://github.com/MobleyLab/alchemical-analysis) script.

## Installation

The package containing `flamel` is called **alchemistry-flamel**. The
latest release can be installed with `pip` or alternatively, install
from source. Both methods are explained below.

### `pip`
*flamel* is available from the Python Package index (PyPi) under the
name **alchemistry-flamel** and can be installed with
```shell
pip install alchemistry-flamel
```
The installed package makes the `flamel` script available.


### From sources
Clone the *flamel* repository https://github.com/alchemistry/flamel
and install with `pip`
```shell
git clone git@github.com:alchemistry/flamel.git
pip install flamel/
```

### Uninstalling
If you want to remove *flamel* after having it installed with `pip`,
run
```shell
pip uninstall alchemistry-flamel
```
to delete `flamel` and its associated files.


## Usage

The analysis can be invoked with the following command

```shell
flamel -a GROMACS -d dhdl_data -f 10 -g -i 50 -j result.csv -m TI,BAR,MBAR -n dE -o out_data -p dhdl -q xvg -r 3 -s 50 -t 298 -v  -w
```

Run ``flamel -h`` to see the full description of the options.

## Output

This script is a wrapper around the 
[ABFE](https://alchemlyb.readthedocs.io/en/latest/workflows/alchemlyb.workflows.ABFE.html#alchemlyb.workflows.ABFE) 
workflow in [alchemlyb](https://github.com/alchemistry/alchemlyb). 
The script will generate the output from ABFE workflow, including 
[O_MBAR.pdf](https://alchemlyb.readthedocs.io/en/latest/visualisation.html#overlap-matrix-of-the-mbar),
[dF_t.pdf](https://alchemlyb.readthedocs.io/en/latest/visualisation.html#df-states-plots-between-different-estimators),
[dF_state.pdf](https://alchemlyb.readthedocs.io/en/latest/visualisation.html#overlap-matrix-of-the-mbar),
[dF_t.pdf](https://alchemlyb.readthedocs.io/en/latest/visualisation.html#forward-and-backward-convergence),
[dhdl_TI.pdf](https://alchemlyb.readthedocs.io/en/latest/visualisation.html#dhdl-plot-of-the-ti).

The script will also generate the `result.csv` and `result.p`, which is a 
pandas DataFrame summarising the results. ::

                      TI  TI_Error    BAR  BAR_Error   MBAR  MBAR_Error
    States 0 -- 1  0.962     0.007  0.956      0.007  0.964       0.006
           1 -- 2  0.567     0.006  0.558      0.006  0.558       0.004
           2 -- 3  0.264     0.005  0.258      0.005  0.254       0.004
           3 -- 4  0.035     0.004  0.035      0.004  0.030       0.003
    Stages fep     1.828     0.014  1.806      0.016  1.807       0.014
           TOTAL   1.828     0.014  1.806      0.011  1.807       0.014

## Name

In the tradition to associate free energy estimations with alchemistry it's 
named after [Nicolas Flamel](https://en.wikipedia.org/wiki/Nicolas_Flamel)

## Copyright

Copyright (c) 2022, the [AUTHORS](./AUTHORS).


## Acknowledgements

@harlor started *flamel* as a replacement for the original
`alchemical-analyis.py` script.

Project template based on the [Computational Molecular Science Python
Cookiecutter](https://github.com/molssi/cookiecutter-cms) version 1.1.



CHANGES
# -*- tab-width: 2; indent-tabs-mode: nil; coding: utf-8 -*-
====================
 flamel CHANGELOG
=====================

The rules for this file:
  * entries are sorted newest-first.
  * summarize sets of changes - don't reproduce every git log comment here.
  * don't ever delete anything.
  * keep the format consistent (79 char width, ISO 8601 date format
    YYYY-MM-DD) and do not use tabs but use spaces for formatting    
  * accompany each entry with github issue/PR number (Issue #xyz)
  * release numbers follow "Semantic Versioning" https://semver.org

------------------------------------------------------------------------------
2022-??-??  xiki-tempula

  * 0.3.2
  
Fixes

- enable logger output for -v (issue #25)


2022-12-06  xiki-tempula, hejamu

  * 0.3.1

Fixes

- Handle 0 as input to the forwrev (Issue #18/ PR #20).


2022-11-09  xiki-tempula, orbeckst

  * 0.3.0

Complete rewrite of the package, it is generally considered
incompatible with previous releases. See Discussion #12
https://github.com/alchemistry/flamel/discussions/12 for background.

Changes

- Removed plugin-based system and changed into a thin CLI to access
  alchemlyb.workflows (see Discussion #12).

Enhancements

- plotting output from ABFE workflow (issue #7)
- pip installable (issue #3)
- supports Python 3.11


2021-09-21  hejamu, schlaicha

  * 0.2.0

Enhancements

- pickle output


Fixes

- corrected unit handling
- reading of GROMACS input files


2018-11-26  harlor

  * 0.1.0

First release (actual release day: 2022-11-06, using @harlor's
original code dating back to 2018).

Features

- plugin-based system
- currently only Gromacs parser and uncorrelation by dH/dl is supported
- reproduces TI, BAR, and MBAR results from alchemical-analysis.py


flamel/flamel.py
"""Provide the primary functions."""

import argparse
import logging
import pathlib
import pickle
import sys

from alchemlyb.workflows import ABFE


def main():
    parser = argparse.ArgumentParser(
        description="""
                    Collect data and estimate free energy differences
                    """,
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "-a",
        "--software",
        dest="software",
        help="Package's name the data files come from: "
        "GROMACS or AMBER. Default: GROMACS.",
        default="GROMACS",
    )
    parser.add_argument(
        "-d",
        "--dir",
        dest="datafile_directory",
        help="Directory in which data files are stored. " "Default: Current directory.",
        default="./",
    )
    parser.add_argument(
        "-f",
        "--forwrev",
        dest="forwrev",
        help="Plot the free energy change as a function of "
        "time in both directions, with the specified "
        "number of points in the time plot. The number "
        "of time points (an integer) must be provided. "
        "Default: 0",
        default=0,
        type=int,
    )
    parser.add_argument(
        "-g",
        "--breakdown",
        dest="breakdown",
        help="Plot the free energy differences evaluated for "
        "each pair of adjacent states for all methods, "
        "including the dH/dlambda curve for TI."
        "Default: True.",
        default=True,
        action="store_true",
    )
    parser.add_argument(
        "-i",
        "--threshold",
        dest="uncorr_threshold",
        help="Proceed with correlated samples if the number of "
        "uncorrelated samples is found to be less than this"
        " number. If 0 is given, the time series analysis "
        "will not be performed at all. Default: 50.",
        default=50,
        type=int,
    )
    parser.add_argument(
        "-j",
        "--resultfilename",
        dest="resultfilename",
        help="custom defined result filename prefix. " "Default: results.csv",
        default="results.csv",
    )

    parser.add_argument(
        "-m",
        "--methods",
        dest="methods",
        help="A comma separated list of the methods to estimate "
        "the free energy with. Default: TI,BAR,MBAR.",
        default="TI,BAR,MBAR",
    )
    parser.add_argument(
        "-n",
        "--uncorr",
        dest="uncorr",
        help="The observable to be used for the "
        "autocorrelation analysis; either 'all' "
        "(obtained as a sum over all energy components) "
        "or 'dE'. In the latter case the energy "
        "differences dE_{i,i+1} (dE_{i,i-1} for the last "
        "lambda) are used.",
        default="dE",
    )
    parser.add_argument(
        "-o",
        "--out",
        dest="output_directory",
        help="Directory in which the output files produced by "
        "this script will be stored. Default: Same as "
        "datafile_directory.",
        default="",
    )
    parser.add_argument(
        "-p",
        "--prefix",
        dest="prefix",
        help="Prefix for datafile sets, i.e.'dhdl' (default).",
        default="dhdl",
    )
    parser.add_argument(
        "-q",
        "--suffix",
        dest="suffix",
        help="Suffix for datafile sets, i.e. 'xvg' (default).",
        default="xvg",
    )
    parser.add_argument(
        "-r",
        "--decimal",
        dest="decimal",
        help="The number of decimal places the free energies "
        "are to be reported with. No worries, this is for "
        "the text output only; the full-precision data "
        "will be stored in 'results.pickle'. Default: 3.",
        default=3,
        type=int,
    )
    parser.add_argument(
        "-s",
        "--skiptime",
        dest="equiltime",
        help="Discard data prior to this specified time as "
        "'equilibration' data. Units picoseconds. "
        "Default: 0 ps.",
        default=0,
        type=float,
    )
    parser.add_argument(
        "-t",
        "--temperature",
        dest="temperature",
        help="Temperature in K. Default: 298 K.",
        default=298.0,
        type=float,
    )
    parser.add_argument(
        "-u",
        "--units",
        dest="units",
        help="Units to report energies: 'kJ/mol', 'kcal/mol', and 'kT'. Default: 'kcal/mol'",
        default="kcal/mol",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="verbose",
        help="Verbose option. Default: False.",
        default=False,
        action="store_true",
    )
    parser.add_argument(
        "-w",
        "--overlap",
        dest="overlap",
        help="Print out and plot the overlap matrix. Default: True.",
        default=True,
        action="store_true",
    )

    args = parser.parse_args(sys.argv[1:])

    # Print the logging to the console
    if args.verbose:
        logging.basicConfig(
            level=logging.INFO,
            handlers=[logging.StreamHandler()],
        )

    if args.output_directory == "":
        out = args.datafile_directory
    else:
        out = args.output_directory
    if args.overlap:
        overlap = "O_MBAR.pdf"
    else:
        overlap = ""

    if args.forwrev > 0:
        forwrev = args.forwrev
    else:
        forwrev = None
    workflow = ABFE(
        T=args.temperature,
        units=args.units,
        software=args.software,
        dir=args.datafile_directory,
        prefix=args.prefix,
        suffix=args.suffix,
        outdirectory=out,
    )
    workflow.run(
        skiptime=args.equiltime,
        uncorr=args.uncorr,
        threshold=args.uncorr_threshold,
        estimators=args.methods.split(","),
        overlap=overlap,
        breakdown=args.breakdown,
        forwrev=forwrev,
    )
    summary = workflow.summary
    pickle.dump(summary, open(pathlib.Path(out) / "result.p", "wb"))
    summary.round(args.decimal).to_csv(pathlib.Path(out) / args.resultfilename)


if __name__ == "__main__":
    main()
